daffodil-runtime1/src/main/scala/org/apache/daffodil/udf/UserDefinedFunctionService.scala

```diff
@@ -0,0 +1,185 @@
+  private val functionProviderLookup: HashMap[NamespaceNameKey, UserDefinedFunctionProvider] = HashMap()
+  private val InitializedUserDefinedFunctionsCache: HashMap[NamespaceNameKey, Option[UdfCallingInfo]] = HashMap()
+  lazy val classUserDefinedFunctionInfo = classOf[UserDefinedFunctionInfo]
+  lazy val classUserDefinedFunction = classOf[UserDefinedFunction]
+  lazy val classJSerializable = classOf[java.io.Serializable]
+  lazy val currentClassPath = Misc.classPath.map(_.toString).mkString("\n")
+
+  val loader: ServiceLoader[UserDefinedFunctionProvider] = ServiceLoader.load(classOf[UserDefinedFunctionProvider])
+
+  loader.asScala.map { provider =>
+    val providerfc = provider.getUserDefinedFunctionClasses
+    lazy val providerClassName = provider.getClass.getName
+
+    if (providerfc == null || providerfc.isEmpty) {
+      val wstr = s"User Defined Function Provider ignored: ${providerClassName}. No User Defined Functions found."
+      log(LogLevel.Warning, wstr)
+    } else {
+      val missingAnnotations = providerfc.filterNot(_.isAnnotationPresent(classUserDefinedFunctionInfo))
+      val nonUserDefinedFunction = providerfc.filterNot(classUserDefinedFunction.isAssignableFrom(_))
+
+      if (!missingAnnotations.isEmpty) {
+        val wstr = s"User Defined Function Provider ignored: ${providerClassName}. Below must be annotated with ${classUserDefinedFunctionInfo.getName}" +
+          s"\n${missingAnnotations.map(_.getName).mkString("\t", "\n\t", "")}"
+        log(LogLevel.Warning, wstr)
+      }
+
+      if (!nonUserDefinedFunction.isEmpty) {
+        val wstr = s"User Defined Function Provider ignored: ${providerClassName}. Below must implement ${classUserDefinedFunction.getName}" +
+          s"\n${nonUserDefinedFunction.map(_.getName).mkString("\t", "\n\t", "")}"
+        log(LogLevel.Warning, wstr)
+      }
+
+      if (missingAnnotations.isEmpty && nonUserDefinedFunction.isEmpty) {
+        providerfc.map { fc =>
+          val fcClassName = fc.getName
+          val fInfo = fc.getDeclaredAnnotation(classUserDefinedFunctionInfo)
+          val fns = fInfo.namespaceURI()
+          val fname = fInfo.name()
+          if (List(fns, fname).exists { p => Option(p).isEmpty || p.trim.isEmpty }) {
+            val wstr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}." +
+              (if (Option(fns).isEmpty || fns.trim.isEmpty) " Annotation namespace field is empty or invalid." else "") +
+              (if (Option(fname).isEmpty || fname.trim.isEmpty) " Annotation namespace field is empty or invalid." else "")
+            log(LogLevel.Warning, wstr)
+          }
+          val key = s"$fns:$fname"
+          if (functionProviderLookup.contains(key)) {
+            val estr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}. Duplicate $key found."
+            log(LogLevel.Error, estr)
```

Comment:
1. Yeah, since this isn't fatal until someone tries to actually use it, a warning seems reasonable.

2. I have no strong opinion one way or the other. Either keep which ever appears first or drop both seems reasonable to me. That said, the logic can get much more complicated if you try to remove dups. For example, say you do the simple thing of just removing the original and ignoring the dup when you see a dup. But imagine there are three dups. When we see the first we add it to the cach. When we see the second, it's a dup so we remove the original and ignore it. Now when we see the third, the original is gone, so we won't realize it's a dup. So the logic needs to be more complex, either rememember which ones are dups, or filter out all dups at the end, or somethign else. Personally, I'm not sure the extra logic is worth it. We'll still output a dup warning, so that should let the user know something is wrong and fix it, or there's a potential of undefined behavior.

Reply:
Ah good point re 2, I'll drop the new dup and leave the original in.

Reply:
Are we talking about detecting duplicate UDFs in different service provider jars on the classpath?

My usual experience of classpath loading is that it is order sensitive, and you are explicitly allowed to have the same thing defined more than once. The first one in classpath order "wins", and this is an intentional way users can do partial overriding of functionality by inserting a new jar ahead on the classpath. 

So if two service providers jars are on the classpath and both provide UDFs and they provide the same UDF definition, then I think the first should "win" and the second and beyond should be ignored, probably silently since it can be an expected ordinary configuration that has this sort of classpath overriding going on. I.e., it's not necessarily anything to warn about.

Reply:
Hmm, we were specifically referring to multiple UDFs with the same name/namespaceUri. The algorithm doesn't really care about the underlying implementation of the UDF as long as the it's considered valid. It does care what name/namespace you assign to it in the @UserDefinedFunctionIdentification annotation. 

So an would be if a provider provides a UDF with a name/namespace that some other UDF (either same or diff provider) already used.

Reply:
The way I'd expect it to work is if there is more than one UDF service provider on the classpath, and the UDF they provide has the same name+namespace, I believe that's not an error or even unusual. The first one just wins and is used, and the second+subsequent are ignored.